op-acceptor/logging/filelogger.go

```diff
@@ -720,3 +751,636 @@ func (s *ConciseSummarySink) Complete(runID string) error {
+	// 4. Add a separator before the error summary section
+	if result.Status == types.TestStatusFail || result.Status == types.TestStatusError {
+		fmt.Fprintf(&content, "\n%s\n", strings.Repeat("-", 80))
+		fmt.Fprintf(&content, "ERROR SUMMARY:\n")
+		fmt.Fprintf(&content, "=============\n\n")
+
+		// Extract critical error information
+		errorInfo := extractErrorInfoFromJSON(result.Stdout)
+
+		if errorInfo.TestName != "" {
+			fmt.Fprintf(&content, "Test:       %s\n", errorInfo.TestName)
+		}
+
+		if errorInfo.ErrorMessage != "" {
+			fmt.Fprintf(&content, "Error:      %s\n", errorInfo.ErrorMessage)
+		}
+
+		if errorInfo.Expected != "" && errorInfo.Actual != "" {
+			fmt.Fprintf(&content, "Expected:   %s\n", errorInfo.Expected)
+			fmt.Fprintf(&content, "Actual:     %s\n", errorInfo.Actual)
+		}
+
+		if errorInfo.Messages != "" {
+			fmt.Fprintf(&content, "Message:    %s\n", errorInfo.Messages)
+		}
+
+		if errorInfo.ErrorTrace != "" {
+			fmt.Fprintf(&content, "\nError Trace:\n%s\n", errorInfo.ErrorTrace)
+		}
+	} else {
+		// For passed tests, a simpler summary at the end
+		fmt.Fprintf(&content, "\n%s\n", strings.Repeat("-", 80))
+		fmt.Fprintf(&content, "RESULT SUMMARY:\n")
+		fmt.Fprintf(&content, "===============\n\n")
+		fmt.Fprintf(&content, "Test passed: %s\n", result.Metadata.FuncName)
+		fmt.Fprintf(&content, "Duration:    %s\n", formatDuration(result.Duration))
+	}
+
+	// Write the content to the file
+	return writer.Write([]byte(content.String()))
+}
+
+// extractPlaintextFromJSON extracts all "Output" fields from JSON to reconstruct the complete plaintext output
+func extractPlaintextFromJSON(jsonOutput string) string {
+	if jsonOutput == "" {
+		return ""
+	}
+
+	var outputBuilder strings.Builder
+	lines := strings.Split(jsonOutput, "\n")
+
+	for _, line := range lines {
+		// Skip empty lines
+		if strings.TrimSpace(line) == "" {
+			continue
+		}
+
+		// Only process JSON lines
+		if !strings.HasPrefix(strings.TrimSpace(line), "{") || !strings.HasSuffix(strings.TrimSpace(line), "}") {
+			continue
+		}
+
+		// Try to parse as JSON
+		var jsonData map[string]interface{}
+		if err := json.Unmarshal([]byte(line), &jsonData); err != nil {
+			continue
+		}
+
+		// Extract information only from output actions
+		action, ok := jsonData["Action"].(string)
+		if !ok || action != "output" {
+			continue
+		}
+
+		// Extract the output text
+		outputText, ok := jsonData["Output"].(string)
+		if !ok || outputText == "" {
+			continue
+		}
+
+		// Add the output text to our builder
+		outputBuilder.WriteString(outputText)
+	}
+
+	return outputBuilder.String()
+}
+
+// ErrorInfo holds extracted error information from test output
+type ErrorInfo struct {
+	TestName     string
+	ErrorMessage string
+	Expected     string
+	Actual       string
+	Messages     string
+	ErrorTrace   string
+}
+
+// extractErrorInfoFromJSON parses JSON test output to extract human-readable error information
+func extractErrorInfoFromJSON(output string) ErrorInfo {
```

Comment:
I think we could consolidate most of the logic of these 2 functions, and generalize it, by having a construct that takes something like an io.Reader as input and produces a stream of json.RawMessage.
Then on the backend of that construct, have handlers that receive those json.RawMessage and extract the relevant information for their purpose (output to reconstruct the pain output, and so on)

Reply:
Done.
I created a JSONOutputParser construct. It then has a method to parse the JSON lines which accepts a handler as a parameter.